[RFC] Deprecate <zephyr/zephyr.h>

gmarull opened the pull request:
As of today, <zephyr/zephyr.h> is 100% equivalent to <zephyr/kernel.h>.
This patch proposes to then include <zephyr/kernel.h> instead of
<zephyr/zephyr.h> since it is clearer that you are including the
Kernel APIs and (probably) nothing else. <zephyr/zephyr.h> sounds like a
catch-all header that may be confusing. Most applications need to
include a bunch of other things to compile, e.g. driver headers or
subsystem headers like BT, logging, etc.

The idea of a catch-all header in Zephyr is likely not feasible
anyway. The reason is that Zephyr is not a library, like it could be for
example `libpython`. Zephyr provides many utilities nowadays: a kernel,
drivers, subsystems, etc. and list will likely grow. A catch-all header
would be massive, difficult to keep up-to-date. It is also likely that
an application will only build a small subset. Note that subsystem-level
headers may use a catch-all approach to make things easier, though.

Resolves https://github.com/zephyrproject-rtos/zephyr/issues/49578

gmarull committed:
includes: prefer <zephyr/kernel.h> over <zephyr/zephyr.h>

As of today <zephyr/zephyr.h> is 100% equivalent to <zephyr/kernel.h>.
This patch proposes to then include <zephyr/kernel.h> instead of
<zephyr/zephyr.h> since it is more clear that you are including the
Kernel APIs and (probably) nothing else. <zephyr/zephyr.h> sounds like a
catch-all header that may be confusing. Most applications need to
include a bunch of other things to compile, e.g. driver headers or
subsystem headers like BT, logging, etc.

The idea of a catch-all header in Zephyr is probably not feasible
anyway. Reason is that Zephyr is not a library, like it could be for
example `libpython`. Zephyr provides many utilities nowadays: a kernel,
drivers, subsystems, etc and things will likely grow. A catch-all header
would be massive, difficult to keep up-to-date. It is also likely that
an application will only build a small subset. Note that subsystem-level
headers may use a catch-all approach to make things easier, though.

NOTE: This patch is **NOT** removing the header, just removing its usage
in-tree. I'd advocate for its deprecation (add a #warning on it), but I
understand many people will have concerns.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

microbuilder commented:
Agree this is clearer.

de-nordic reviewed (dismissed):
Subsystems mcumgr and VFS are OK.

stephanosio reviewed (dismissed):
Looks reasonable.

After the introduction of the `zephyr/` namespace, I have always thought `zephyr/zephyr.h` looked silly (if any, it should be just `zephyr.h` because that in its own is unique; but, that is for another day ...).

carlocaione reviewed (dismissed):
LGTM for the files I own.

nashif commented:
> NOTE: This patch is **NOT** removing the header, just removing its usage
> in-tree. I'd advocate for its deprecation (add a #warning on it), but I
> understand, many people will have concerns.

if it is not deprecated and officially documented as not to be used anymore, then it will be re-introduced and will creep slowly into the code base and reviewers will have to chase after PRs including it and the whole deal.

This change IMO needs to be supported by a clear plan: remove and deprecate and document, and then set a goal to completely remove. And it is better to do that now than later, say if we want to drop this for LTS3, I would start now.

gmarull committed:
manifest: update hal_ti

Update hal to remove usage of <zephyr/zephyr.h>.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
includes: deprecate <zephyr/zephyr.h>

Deprecate <zephyr/zephyr.h> in favor of <zephyr/kernel.h> since they are
equivalents (zephyr/zephyr.h is just a shim). Do not warn about
deprecation if already using the deprecated CONFIG_LEGACY_INCLUDE_PATH
option.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: release-notes: add notes about <zephyr/zephyr.h> deprecation

Inform about the <zephyr/zephyr.h> deprecation, add details on what
users need to do.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
> > NOTE: This patch is **NOT** removing the header, just removing its usage
> > in-tree. I'd advocate for its deprecation (add a #warning on it), but I
> > understand, many people will have concerns.
> 
> if it is not deprecated and officially documented as not to be used anymore, then it will be re-introduced and will creep slowly into the code base and reviewers will have to chase after PRs including it and the whole deal.
> 
> This change IMO needs to be supported by a clear plan: remove usage, deprecate and document, and then set a goal to completely remove. And it is better to do that now than later, say if we want to drop this for LTS3, I would start now.

updated PR with deprecation

str4t0m commented on samples/net/lldp/src/main.c:
```diff
@@ -7,9 +7,9 @@
 
-#include <zephyr/zephyr.h>
+#include <zephyr/kernel.h>
 
-#include <zephyr/zephyr.h>
+#include <zephyr/kernel.h>
```
It is included twice in this file, could remove one.

gmarull replied:
fixed

gmarull committed:
samples: net: lldp: remove duplicate include

<zephyr/zephyr.h> was included twice.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

andyross reviewed (dismissed):
No complaints here.  Someone merge it before any of us get brave enough to form opinions.

mbolivar-nordic commented:
> No complaints here. Someone merge it before any of us get brave enough to form opinions.

I want to agree with you, but I think we ought to treat this via the 'treewide changes' change control process we are slowly converging on in https://github.com/zephyrproject-rtos/zephyr/issues/48886.

nashif commented:
DNM per https://github.com/zephyrproject-rtos/zephyr/pull/49510#issuecomment-1227793723

carlescufi commented:
API meeting:

- Mostly in favor of the change in general
- `LEGACY_INCLUDE_PATH` could be used to enable/disable the whole transition:
   - Switch from `<zephyr.h>` to `<zephyr/kernel.h>`
- There's a script to help with the transition

carlescufi commented on include/zephyr/zephyr.h:
```diff
@@ -9,4 +9,6 @@
 
 #include <zephyr/kernel.h>
 
+#warning "<zephyr/zephyr.h> is deprecated, include <zephyr/kernel.h> instead"
```
This should say "Set CONFIG_LEGACY_INCLUDE_PATH=y" to disable the warning

carlescufi replied:
@gmarull @henrikbrixandersen @nashif 

@rodrigopex suggests using `#message` instead of `#warning`, this might bee a good intermediate solution.

gmarull replied:
is pragma message supported by all compilers?

stephanosio replied:
> is pragma message supported by all compilers?

Neither `#warning` nor `#pragma message` are part of the ISO C standard, but `#warning` is so widely supported that we use it without any checks.

Given that `#pragma message` is also universally supported, I see no reason why it should be treated differently.

carlescufi commented on include/zephyr/zephyr.h:
```diff
@@ -9,4 +9,6 @@
 
 #include <zephyr/kernel.h>
 
+#warning "<zephyr/zephyr.h> is deprecated, include <zephyr/kernel.h> instead"
```
```
#if !defined(CONFIG_LEGACY_INCLUDE_PATH)
#warning "<zephyr/zephyr.h> is deprecated, include <zephyr/kernel.h> instead"
#endif
```

gmarull replied:
adopted this solution, `#pragma message` requires extra infrastructure upstream to prevent its usage. Also, if we want to introduce soft-deprecation I think we need to consider all kind of deprecations (e.g. functions, macros).

carlescufi reviewed (changes requested):
API meeting changes

rodrigopex commented:
@gmarull maybe #pragma message "this is a message" could solve the #warning issue. Thus, it only sends a note, not a warning, during the compilation.

gmarull committed:
scripts: migrate_includes: migrate zephyr.h to kernel.h

The script will now automatically migrate all <zephyr.h> or
<zephyr/zephyr.h> (for already migrated projects) to <zephyr/kernel.h>.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

stephanosio commented:
> DNM per [#49510 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/49510#issuecomment-1227793723)

@nashif Do you mean we should go through the process documented in #48886 for this PR? (though, the process has not been finalised yet).

gmarull commented:
@stephanosio @carlescufi should now be ready

carlescufi commented:
> > No complaints here. Someone merge it before any of us get brave enough to form opinions.
> 
> I want to agree with you, but I think we ought to treat this via the 'treewide changes' change control process we are slowly converging on in #48886.

We don´t seem to have converged to anything actionable yet though @mbolivar-nordic. I think we should not enforce a process that is still under development.

gmarull commented:
Please re-review again (rebased due to conflicts). @fabiobaltieri @carlescufi please merge as soon as green and approved to avoid yet another round of rebase/re-approvals nonsense.

carlescufi commented:
> > > No complaints here. Someone merge it before any of us get brave enough to form opinions.
> > 
> > 
> > I want to agree with you, but I think we ought to treat this via the 'treewide changes' change control process we are slowly converging on in #48886.
> 
> We don´t seem to have converged to anything actionable yet though @mbolivar-nordic and @nashif . I think we should not enforce a process that is still under development.

Actually @gmarull has followed the procedure described in the working document, so we can merge this safely.
